### PR TITLE
`Sendable` support

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes [CSDK-379] and https://github.com/RevenueCat/purchases-ios/issues/1041#issuecomment-1206445939

[CSDK-379]: https://revenuecats.atlassian.net/browse/CSDK-379?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### Main changes:
- `Sendable` conformances to all types that are thread-safe
- `@unchecked Sendable` for types that the compiler can't enforce, but with documentation as to why
- Made classes that aren't mocked `final` (note that this doesn't change the API, because none of these were `open` to begin with)
- Made some classes thread-safe that weren't (like `DeviceCache`)

For the non-`final` `class`es, I've actually managed to compile the SDK with all of them as `final` and `Sendable`, to make sure that making them `@unchecked Sendable` didn't hide any other issues.

### Future improvements:
- https://twitter.com/nachosoto/status/1557139992056500224?s=21&t=arWdvEzTIFANBvwqQ0vPiA
- https://twitter.com/nachosoto/status/1557141944777592832?s=21&t=arWdvEzTIFANBvwqQ0vPiA
- https://twitter.com/nachosoto/status/1557143374922059776?s=21&t=arWdvEzTIFANBvwqQ0vPiA

### Depends on:
- #1794
- #1804
- #1806
- #1807
- #1808
- #1813
- #1822
- #1823
- #1824
- #1825
- #1826

### CODE REVIEW

File: RevenueCat.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -3168,6 +3168,7 @@
 				SDKROOT = iphoneos;
 				SUPPORTS_MACCATALYST = YES;
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
+				SWIFT_STRICT_CONCURRENCY = targeted;
```
Comment (NachoSoto) [inline RevenueCat.xcodeproj/project.pbxproj:3171]: We'll see if we keep this or not.

**Comment by NachoSoto** [comment]
This is almost ready, but it will be a lot easier to review once all those dependent PRs are merged.

File: RevenueCat.xcodeproj/project.pbxproj (hunk 1 above)
Comment (NachoSoto) [reply RevenueCat.xcodeproj/project.pbxproj:3188]: I think it's not a bad idea to keep this so we ensure we maintain this level of correctness prior to Swift 6.0

**Comment by NachoSoto** [comment]
This is ready to review @RevenueCat/core-sdk

**Comment by NachoSoto** [comment]
Bump @RevenueCat/core-sdk

**Review by taquitos — APPROVED** [review]
😮‍💨
Nice work.

File: Sources/Purchasing/Purchases/Purchases.swift [hunk 2]
```diff
@@ -1628,6 +1629,16 @@ public extension Purchases {
...
 }
 
+// @unchecked because:
+// - It contains `NotificationCenter`, which isn't thread-safe as of Swift 5.7.
+// - It has a mutable `privateDelegate` (this isn't actually thread-safe!)
+// - It has a mutable `customerInfoObservationDisposable` because it's late-initialized in the constructor
+//
+// One could argue this warrants making this class non-Sendable, but the annotation allows its usage in
+// async contexts in a much more simple way without errors like:
+// "Capture of 'self' with non-sendable type 'Purchases' in a `@Sendable` closure"
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/Purchases/Purchases.swift:1639]: I agree, which would help continue making `Purchases` smaller

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift [hunk 3]
```diff
@@ -652,14 +665,24 @@ private extension PurchasesOrchestrator {
...
+            self.operationDispatcher.dispatchOnMainActor {
+                completion(
+                    nil,
+                    nil,
+                    ErrorUtils.storeProblemError(
+                        withMessage: Strings.purchase.could_not_purchase_product_id_not_found.description
+                    ),
+                    false
+                )
+            }
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/Purchases/PurchasesOrchestrator.swift:677]: Can you clarify which style you mean?

File: Sources/Purchasing/TrialOrIntroPriceEligibilityChecker.swift [hunk 4]
```diff
@@ -142,11 +142,9 @@ fileprivate extension TrialOrIntroPriceEligibilityChecker {
                                              completion: completion)
                     return
                 }
-                var convertedEligibility: [String: IntroEligibility] = [:]
-                for (key, value) in receivedEligibility {
-                    let introEligibility = IntroEligibility(eligibilityStatus: value)
-                    convertedEligibility[key] = introEligibility
-                }
+
+                let convertedEligibility = receivedEligibility.mapValues(IntroEligibility.init)
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:146]: I hope that's a good face palm and not a bad one haha

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (taquitos) [reply Sources/Purchasing/Purchases/PurchasesOrchestrator.swift:677]: Going from single-line to multi-line completion call.

File: Sources/Purchasing/TrialOrIntroPriceEligibilityChecker.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (taquitos) [reply Sources/Purchasing/TrialOrIntroPriceEligibilityChecker.swift:146]: It's a good one 😄

File: Sources/Purchasing/Purchases/PurchasesOrchestrator.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Purchasing/Purchases/PurchasesOrchestrator.swift:677]: I needed it to avoid the long line I think after adding the `dispatchOnMainActor` :/

**Comment by NachoSoto** [comment]
Merging this unless somebody else has any concerns or wants more time to look at it @RevenueCat/core-sdk

**Review by aboedo — APPROVED** [review]

File: Sources/Misc/OperationDispatcher.swift [hunk 5]
```diff
@@ -16,27 +16,52 @@ import Foundation
...
+            self.workerQueue.async(execute: block)
+        }
+    }
+
+}
+
+extension OperationDispatcher {
+
+    static func dispatchOnMainActor(_ block: @MainActor @escaping @Sendable () -> Void) {
+        if #available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *) {
```
Comment (aboedo) [inline Sources/Misc/OperationDispatcher.swift:51]: this is cool

File: Sources/Misc/OperationDispatcher.swift [hunk 6]
```diff
@@ -16,27 +16,52 @@ import Foundation
...
+            _ = Task {
+                await MainActor.run {
+                    block()
+                }
+            }
         } else {
-            workerQueue.async(execute: block)
+            DispatchQueue.main.async(execute: block)
```
Comment (aboedo) [inline Sources/Misc/OperationDispatcher.swift:58]: should this be `self.mainQueue`?

File: Sources/Misc/OperationDispatcher.swift [hunk 7]
```diff
@@ -16,27 +16,52 @@ import Foundation
...
-    private let mainQueue = DispatchQueue.main
-    private let workerQueue = DispatchQueue(label: "OperationDispatcherWorkerQueue")
+    private let mainQueue: DispatchQueue = .main
+    private let workerQueue: DispatchQueue = .init(label: "OperationDispatcherWorkerQueue")
     private let maxJitterInSeconds: Double = 5
 
     static let `default`: OperationDispatcher = .init()
 
-    func dispatchOnMainThread(_ block: @escaping () -> Void) {
+    func dispatchOnMainThread(_ block: @escaping @Sendable () -> Void) {
```
Comment (aboedo) [inline Sources/Misc/OperationDispatcher.swift:25]: Should this be made private now? Is there any scenario where we should choose to go with main thread and not main actor even when Swift Concurrency is available?

File: Sources/Misc/OperationDispatcher.swift (hunk 6 above)
Comment (NachoSoto) [reply Sources/Misc/OperationDispatcher.swift:58]: Oh yeah good call.

File: Sources/Misc/OperationDispatcher.swift (hunk 7 above)
Comment (NachoSoto) [reply Sources/Misc/OperationDispatcher.swift:25]: Eventually we can get rid of this, but that seems like a bigger change. For existing code that does not use the new Swift "structured" concurrency, we might not want to change the semantics if we don't need to.

There's still a few places using `dispatchOnMainThread` that we might not want to change until we use `@MainActor` (or other custom `actor`s) in more places.

Comment (NachoSoto) [reply Sources/Misc/OperationDispatcher.swift:25]: Let men now what you think, we can migrate more of these to `dispatchOnMainActor` in a following PR.

File: Sources/Misc/OperationDispatcher.swift (hunk 6 above)
Comment (aboedo) [reply Sources/Misc/OperationDispatcher.swift:58]: ohhh right, that makes sense. Could you add a comment for the future?

File: Sources/Misc/OperationDispatcher.swift (hunk 7 above)
Comment (aboedo) [reply Sources/Misc/OperationDispatcher.swift:25]: 👍

**Pull request merged**